lib/timeline-window.js

```diff
@@ -0,0 +1,392 @@
+
+/**
+ * @private
+ */
+var debuglog = DEBUG ? console.log.bind(console) : function() {};
+
+/**
+ * Construct a TimelineWindow.
+ *
+ * <p>This abstracts the separate timelines in a Matrix {@link
+ * module:models/room~Room|Room} into a single iterable thing. It keeps track of
+ * the start and endpoints of the window, which can be advanced with the help
+ * of pagination requests.
+ *
+ * <p>Before the window is useful, it must be initialised by calling {@link
+ * module:timeline-window~TimelineWindow#load|load}.
+ *
+ * <p>Note that the window will not automatically extend itself when new events
+ * are received from /sync; you should arrange to call {@link
+ * module:timeline-window~TimelineWindow#paginate|paginate} on {@link
+ * module:client~MatrixClient.event:"Room.timeline"|Room.timeline} events.
+ *
+ * @param {MatrixClient} client   MatrixClient to be used for context/pagination
+ *   requests.
+ *
+ * @param {Room} room  The room to track
+ *
+ * @param {Object} [opts] Configuration options for this window
+ *
+ * @param {number} [opts.windowLimit = 1000] maximum number of events to keep
+ *    in the window. If more events are retrieved via pagination requests,
+ *    excess events will be dropped from the other end of the window.
+ *
+ * @constructor
+ */
+function TimelineWindow(client, room, opts) {
+    opts = opts || {};
+    this._client = client;
+    this._room = room;
+
+    // _start.index is inclusive; _end.index is exclusive.
+    this._start = null;
+    this._end = null;
+
+    this._eventCount = 0;
+    this._windowLimit = opts.windowLimit || 1000;
+}
+
+/**
+ * Initialise the window to point at a given event, or the live timeline
+ *
+ * @param {string} [initialEventId]   If given, the window will contain the
+ *    given event
+ * @param {number} [initialWindowSize = 20]   Size of the initial window
+ *
+ * @return {module:client.Promise}
+ */
+TimelineWindow.prototype.load = function(initialEventId, initialWindowSize) {
+    var self = this;
+    initialWindowSize = initialWindowSize || 20;
+
+    var prom;
+    if (initialEventId) {
+        debuglog("TimelineWindow: initialising for event " + initialEventId);
+        prom = this._client.getEventTimeline(this._room, initialEventId)
+            .then(function(tl) {
+                // make sure that our window includes the event
+                for (var i = 0; i < tl.getEvents().length; i++) {
+                    if (tl.getEvents()[i].getId() == initialEventId) {
+                        return [tl, i];
+                    }
+                }
+                throw new Error("getEventTimeline result didn't include requested event");
+            });
+    } else {
+        debuglog("TimelineWindow: initialising with live timeline");
+
+        // start with the most recent events
+        var tl = this._room.getLiveTimeline();
+        prom = q([tl, tl.getEvents().length]);
+    }
+
+    prom = prom.then(function(v) {
+        var tl = v[0], eventIndex = v[1];
```

Comment:
Better var names please. I would prefer it if you passed the timeline and and the index as an object with descriptive key names rather than an array with magical ordering:
```javascript
{
  timeline: ...
  index: 5
}
```